cpu/rp2040/cpu.c

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (C) 2021 Otto-von-Guericke Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_rp2040
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the CPU initialization
+ *
+ * @author      Fabian Hüßler <fabian.huessler@ovgu.de>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "macros/units.h"
+#include "periph/init.h"
+#include "periph_cpu.h"
+#include "io_reg.h"
+#include "stdio_base.h"
+#include "vendor/RP2040.h"
+
+static void _cpu_reset(void)
+{
+    /* 2.14 subsystem resets */
+    uint32_t rst;
+    /* Reset hardware components except for critical ones */
+    rst = RESETS_RESET_MASK &
+          ~(RESETS_RESET_usbctrl_Msk    |
+            RESETS_RESET_syscfg_Msk     |
+            RESETS_RESET_pll_usb_Msk    |
+            RESETS_RESET_pll_sys_Msk    |
+            RESETS_RESET_pads_qspi_Msk  |
+            RESETS_RESET_io_qspi_Msk);
+    periph_reset(rst);
+    /* Assert that reset has completed except for those components which
+       are not clocked by clk_ref or clk_sys */
+    rst = RESETS_RESET_MASK &
+          ~(RESETS_RESET_usbctrl_Msk    |
+            RESETS_RESET_uart1_Msk      |
+            RESETS_RESET_uart0_Msk      |
+            RESETS_RESET_spi1_Msk       |
+            RESETS_RESET_spi0_Msk       |
+            RESETS_RESET_rtc_Msk        |
+            RESETS_RESET_adc_Msk);
+    periph_reset_done(rst);
+
+    /* power the reference clock from its default source: the ROSC */
+    clock_ref_configure_source(MHZ(12), MHZ(12), CLOCKS_CLK_REF_CTRL_SRC_rosc_clksrc_ph);
+    /* power the system clock from its default source: the reference clock */
+    clock_sys_configure_source(MHZ(12), MHZ(12), CLOCKS_CLK_SYS_CTRL_SRC_clk_ref);
+    /* start XOSC running at 12 MHz */
+    xosc_start(MHZ(12));
+    /* reset system PLL */
+    pll_reset_sys();
+    /* make the system PLL output a 125 MHz frequency from the 12 MHz XOSC */
+    pll_start_sys(MHZ(12), 1, 125, 6, 2);
```

Comment:
PLL configuration values:
`12 MHz` input frequency from the `XTAL`,
divided by  `1`,
scaled with `125` by the `PLL`,
post-divided by `(6 * 2)` ,
makes `125` MHz.

Reply:
The settings are taken from the SDK.

Reply:
I moved the numbers to defines in `periph_cpu.h` and allowed to override those with defines. (E.g. up to 133 MHz is officially allowed, and some people clock their silicon with over 200 MHz, apparently to some success. So it might make sense to ease this.)